File: courses/01-objectives-outcomes-requirements/insert-synthesis.mjs
// Postprocess the existing pipeline's HTML; this does not build reference cards.
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
const dir = path.dirname(fileURLToPath(import.meta.url));
const report = path.join(dir, '../01-objectives-outcomes-requirements.html');
const data = JSON.parse(await fs.readFile(path.join(dir, 'references.json'), 'utf8'));
const synthesis = JSON.parse(await fs.readFile(path.join(dir, 'synthesis.json'), 'utf8'));
const refs = new Map(data.references.map(r => [r.id, r]));
const esc = s => String(s).replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
function cite(ids) {
  if (!ids?.length) throw Error('Every synthesis entry needs evidence');
  return ids.map(id => { const r=refs.get(id); if (!r) throw Error(`Unknown reference ${id}`); return `<a class="evidence" href="#ref-${encodeURIComponent(id)}">${esc(r.app)} · ${esc(id)}</a>`; }).join('; ');
}
const list = entries => `<ul>${entries.map(e=>`<li>${esc(e.text)} <span class="citations">${cite(e.refs)}</span></li>`).join('')}</ul>`;
const block = `<section id="synthesis" aria-label="Research synthesis">
<p class="eyebrow">Research synthesis · ${data.references.length} references · ${new Set(data.references.map(r=>r.app)).size} products</p>
<p>${esc(synthesis.scope)}</p>
<h3>Top examples</h3><ol class="top-examples">${synthesis.top.map(e=>{ const r=refs.get(e.ref); if(!r)throw Error('Missing top ref'); return `<li><strong>${esc(r.app)}</strong> <span class="badge">${esc(r.source)}</span><p>${esc(e.why)} ${cite([e.ref])}</p></li>`; }).join('')}</ol>
<h3>Pattern comparison</h3><p class="muted">${esc(synthesis.legend)}</p>
<div class="comparison-scroll" role="region" aria-label="Product comparison across six questions" tabindex="0"><table><thead><tr><th scope="col">Product</th>${data.questions.map(q=>`<th scope="col"><a href="#question-${Number(q.id.slice(1))-1}">${esc(q.id)}</a><br>${esc(q.title)}</th>`).join('')}</tr></thead><tbody>${synthesis.comparison.map(row=>`<tr><th scope="row">${esc(row.app)}</th>${row.cells.map(cell=>`<td>${esc(cell.text)}${cell.refs?.length?`<br><small>${cite(cell.refs)}</small>`:''}</td>`).join('')}</tr>`).join('')}</tbody></table></div>
<h3>What the best apps do differently</h3>${list(synthesis.differences)}
<h3>Best practices top companies converge on</h3><p class="muted">Recommendations inferred from the cited examples; these are design proposals for Essentials 2.0.</p>${list(synthesis.practices)}
<h3>Open questions for Essentials 2.0</h3>${list(synthesis.openQuestions)}
</section>`;
let html=await fs.readFile(report,'utf8');
if(html.includes('id="synthesis"'))throw Error('Rebuild base report before inserting synthesis');
let index=0;
html=html.replace(/<article class="card"/g,()=>`<article id="ref-${encodeURIComponent(data.references[index++].id)}" class="card"`);
if(index!==data.references.length)throw Error('Card/reference count mismatch');
let bodyIndex=0;
html=html.replace(/<div class="card-body">/g,()=>`<div class="card-body"><div class="ref-id">${esc(data.references[bodyIndex++].id)}</div>`);
html=html.replace('</header>', `</header>\n${block}`);
html=html.replace('</style>', `
.ref-id{font-size:10px;color:var(--muted);margin-bottom:12px;overflow-wrap:anywhere}#synthesis{margin:28px 0;padding:24px;background:var(--surface);border:1px solid var(--border);border-radius:8px}#synthesis h3{margin-top:28px}#synthesis li{margin:12px 0;max-width:1100px}#synthesis .citations{display:block;font-size:12px;color:var(--muted)}#synthesis .evidence{overflow-wrap:anywhere}.top-examples{padding-left:24px;columns:2;column-gap:40px}.top-examples li{break-inside:avoid}.top-examples p{margin:6px 0;font-size:14px}.top-examples .evidence{display:block;font-size:11px;color:var(--muted)}.comparison-scroll{overflow-x:auto;margin:16px 0}table{border-collapse:collapse;font-size:12px;min-width:1100px;width:100%}th,td{padding:12px 10px;border:1px solid var(--border);vertical-align:top;text-align:left}thead th{background:var(--subtle);font-weight:600}tbody th{min-width:100px}td{min-width:160px}td small{font-size:10px;color:var(--muted)}article:target{outline:3px solid var(--focus);outline-offset:4px}article{scroll-margin-top:24px}@media(max-width:700px){#synthesis{padding:16px}.top-examples{columns:1}}\n</style>`);
await fs.writeFile(report,html);
console.log(`Synthesis inserted; ${index} reference anchors; ${(Buffer.byteLength(html)/1024/1024).toFixed(2)} MiB`);

File: courses/08-recordings-course-reports/insert-synthesis.py
"""Postprocess the existing shared builder output; does not replace the builder."""
import json, re, html
from pathlib import Path
p=Path(__file__).resolve().parent
report=p.parent/'08-recordings-course-reports.html'
d=json.loads((p/'references.json').read_text())
s=json.loads((p/'synthesis.json').read_text())
refs={r['id']:r for r in d['references']}
def esc(x): return html.escape(str(x))
def cite(ids):
 return ' '.join(f'<a href="#{esc(i)}">[{esc(refs[i]["app"])} · {esc(i)}]</a>' for i in ids)
def items(xs):
 return '<ul>'+''.join('<li>'+esc(x['text'])+' '+cite(x.get('refs',[]))+'</li>' for x in xs)+'</ul>'
block='<section id="synthesis" aria-label="Research synthesis"><h3>Top examples</h3>'+items(s['top'])
block+='<h3>Pattern comparison</h3><p>'+esc(s['legend'])+'</p><div class="comparison"><table><thead><tr><th>Product</th>'+''.join('<th title="'+esc(q['title'])+'">'+q['id']+'</th>' for q in d['questions'])+'</tr></thead><tbody>'
for row in s['comparison']:
 block+='<tr><th>'+esc(row['app'])+'<br>'+' '.join(f'<a href="#{esc(i)}" title="{esc(i)}">[{n+1}]</a>' for n,i in enumerate(row.get('refs',[])))+'</th>'+''.join('<td>'+esc(cell)+'</td>' for cell in row['cells'])+'</tr>'
block+='</tbody></table></div>'
for key,title in [('differences','What the best apps do differently'),('practices','Best practices top companies converge on'),('open','Open questions for Essentials 2.0')]:
 block+='<h3>'+title+'</h3>'+items(s[key])
block+='<p class="muted">'+esc(s['limitations'])+'</p></section>'
h=report.read_text()
h=h.replace('</header>','</header>'+block,1)
css='''#synthesis{margin:28px 0;padding:22px;background:var(--surface);border:1px solid var(--border);border-radius:8px}#synthesis h3{margin-top:24px}#synthesis li{margin:10px 0;max-width:1200px}#synthesis a{overflow-wrap:anywhere;font-size:12px}.comparison{overflow-x:auto}table{border-collapse:collapse;min-width:1050px;width:100%;font-size:13px}th,td{padding:10px;vertical-align:top;border:1px solid var(--border);text-align:left}thead{background:var(--subtle)}table{table-layout:fixed}th:first-child{width:150px}.reference-id{display:block;color:var(--muted);font-size:10px;margin-top:10px;overflow-wrap:anywhere}.subsection-label{margin-top:36px;font-size:13px;text-transform:uppercase;letter-spacing:.07em;color:var(--muted)}'''
h=h.replace('</style>',css+'</style>',1)
# Shared builder renders references in question order. Add stable anchors and IDs.
ordered=[r for q in d['questions'] for r in d['references'] if r['question']==q['id']]
it=iter(ordered)
h=re.sub(r'<article class="card"',lambda m:'<article id="'+esc(next(it)['id'])+'" class="card"',h)
for r in ordered:
 pattern='id="'+esc(r['id'])+'" class="card"'
 start=h.index(pattern);pos=h.index('<div class="card-body">',start)+len('<div class="card-body">')
 h=h[:pos]+'<span class="reference-id">'+esc(r['id'])+'</span>'+h[pos:]
h=h.replace('<section class="question-section" data-question="0"','<div class="subsection-label">Part A · Recording organisation, access and playback</div><section class="question-section" data-question="0"',1)
h=h.replace('<section class="question-section" data-question="3"','<div class="subsection-label">Part B · Course reporting and learner results</div><section class="question-section" data-question="3"',1)
report.write_text(h)
print('Synthesis inserted; six question H2s preserved.')
